fccodegen/_generator.py

```diff
@@ -158,39 +273,86 @@ def derivative_name(deriv):
+    state_info = []
+    for i, state in enumerate(get_state_symbols(graph)):
+        #TODO: Update cellmlmanip so this graph.nodes goes away
+        initial_value = float(graph.nodes[state]['initial_value'])
+        state_info.append({
             'index': i,
             'var_name': symbol_name(state),
             'deriv_name': derivative_name(state),
+            'initial_value': initial_value,
+        })
+
+    # Create parameter information dicts
+    parameter_info = []
+    for i, parameter in enumerate(parameters):
+        symbol = get_symbol_by_cmeta_id(graph, parameter)
+        parameter_info.append({
+            'index': i,
+            'cmeta_id': parameter,
+            'var_name': symbol_name(symbol),
+            'initial_value': get_value(graph, symbol),
+        })
+
+    # Create map of parameter symbols to their indices
+    parameter_symbols = {}
+    for i, parameter in enumerate(parameters):
+        symbol = get_symbol_by_cmeta_id(graph, parameter)
+        parameter_symbols[symbol] = i
+
+    # Create output information dicts
+    # Each output is associated either with a symbol or a parameter.
+    output_info = []
+    for i, output in enumerate(outputs):
+        symbol = get_symbol_by_cmeta_id(graph, output)
+        output_info.append({
+            'index': i,
+            'cmeta_id': output,
+            'var_name': symbol_name(symbol),
+            'parameter_index': parameter_symbols.get(symbol, None),
         })
 
-    # Create the RHS equations
+    # Create RHS equation information dicts
     rhs_equations = []
-    for eq in eqs:
-        lhs, rhs = eq.lhs, eq.rhs
-        rhs_equations.append(p.doprint(lhs) + ' = ' + p.doprint(rhs))
+    for eq in get_equations_for(graph, get_derivative_symbols(graph)):
+        rhs_equations.append({
+            'lhs': printer.doprint(eq.lhs),
+            'rhs': printer.doprint(eq.rhs),
+            'parameter_index': parameter_symbols.get(eq.lhs, None),
+        })
+
+    # Create output equation information dicts
+    output_equations = []
+    output_symbols = [get_symbol_by_cmeta_id(graph, x) for x in outputs]
+    for eq in get_equations_for(graph, output_symbols):
+        output_equations.append({
+            'lhs': printer.doprint(eq.lhs),
+            'rhs': printer.doprint(eq.rhs),
+            'parameter_index': parameter_symbols.get(eq.lhs, None),
+        })
 
     # Generate model
     template = load_template('wl', 'weblab_model.pyx')
     with open(path, 'w') as f:
         f.write(template.render({
-            'date': time.strftime('%Y-%m-%d %H:%M:%S'),
-            'generator': cg.version(formatted=True),
-            'name': model.name,
+            'class_name': class_name,
+            'free_variable': free_name,
+            'generation_date': time.strftime('%Y-%m-%d %H:%M:%S'),
+            'generator_name': cg.version(formatted=True),
+            'model_name': model.name,
+            'n_parameters': len(parameter_info),
```

Comment:
You can get this from the `parameter_info` variable within the template using the `length` [filter](http://jinja.pocoo.org/docs/2.10/templates/#filters) - no need to pass as a separate variable.

Reply:
I think this is cleaner, no?

Reply:
I disagree :) It gives more scope for things to get out of sync (small scope, admittedly, but why pass a new variable you don't need?)